src/Passes/Source/Apps/Qat/Qat.cpp

```diff
@@ -31,35 +64,33 @@ int main(int argc, char** argv)
 {
     try
     {
-        // Parsing commmandline arguments
-        Settings settings{
-            {{"debug", "false"},
-             {"generate", "false"},
-             {"validate", "false"},
-             {"profile", "baseProfile"},
-             {"S", "false"},
-             {"O0", "false"},
-             {"O1", "false"},
-             {"O2", "false"},
-             {"O3", "false"},
-             {"verify-module", "false"}}};
-
-        ParameterParser parser(settings);
-        parser.addFlag("debug");
-        parser.addFlag("generate");
-        parser.addFlag("validate");
-        parser.addFlag("verify-module");
-        parser.addFlag("S");
-        parser.addFlag("O0");
-        parser.addFlag("O1");
-        parser.addFlag("O2");
-        parser.addFlag("O3");
+        auto profile = std::make_shared<DefaultProfileGenerator>();
+
+        ConfigurationManager& configuration_manager = profile->configurationManager();
+        configuration_manager.addConfig<QatConfig>();
+        configuration_manager.addConfig<FactoryConfiguration>();
 
+        // Parsing command line arguments
+        ParameterParser parser;
+        configuration_manager.setupArguments(parser);
         parser.parseArgs(argc, argv);
+        configuration_manager.configure(parser);
+
+        // Getting the main configuration
+        auto const& config = configuration_manager.get<QatConfig>();
+
+        // In case we debug, we also print the settings to allow provide a full
+        // picture of what is going.
+        if (config.dumpConfig())
+        {
+            configuration_manager.printConfiguration();
+        }
 
         if (parser.arguments().empty())
         {
             std::cerr << "Usage: " << argv[0] << " [options] filename" << std::endl;
+            configuration_manager.printHelp();
```

Comment:
Arguably only errors should go to the error stream which is what is happening here. I think it would make more sense to change `llvm::errs()` to `llvm::outs()` the places where it is actually not errors. What are your thoughts?

Reply:
I agree to limit the error stream to be only errors. I think we need to carefully examine and probably replace all statements that write to one of the streams to accumulate and handle that properly in the service. @idavis If I remember correct we briefly touched on this today - you may have some suggestions off the top of your head (for a next PR)?

Reply:
I am happy to discuss this, but think it is out of scope for this PR.

Reply:
Agreed.